File: app/src/main/java/com/gnoemes/shikimoriapp/domain/search/SearchQueryBuilder.java
package com.gnoemes.shikimoriapp.domain.search;

import android.support.annotation.Nullable;

import com.gnoemes.shikimoriapp.entity.search.domain.FilterItem;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;

import io.reactivex.Single;

public interface SearchQueryBuilder {

    Single<Map<String, String>> createQueryFromFilters(Map<String, List<FilterItem>> filters, int page, int limit);

    Single<Map<String, String>> createQueryFromIds(LinkedHashSet<Long> ids, @Nullable String searchQuery, int page, int limit);
}

File: app/src/main/java/com/gnoemes/shikimoriapp/domain/search/SearchQueryBuilder.kt
package com.gnoemes.shikimoriapp.domain.search

import com.gnoemes.shikimoriapp.entity.main.domain.Constants
import com.gnoemes.shikimoriapp.entity.search.domain.FilterItem
import io.reactivex.Single

interface SearchQueryBuilder {

    fun createQueryFromFilters(filters: Map<String, MutableList<FilterItem>>?, page: Int, limit: Int): Single<Map<String, String>>

    fun createQueryFromIds(ids: MutableCollection<Long>, searchQuery: String? = null, page: Int = 1, limit: Int = Constants.MAX_LIMIT): Single<Map<String, String>>

    fun createMyListQueryFromIds(ids: MutableCollection<Long>, searchQuery: String? = null, page: Int = 1, limit: Int = Constants.MAX_LIMIT): Single<Map<String, String>>
}